src/kv/lazy-store.ts

```diff
@@ -0,0 +1,235 @@
+import {RWLock} from '../rw-lock';
+import type {Read, Store, Value, Write} from './store';
+import {deleteSentinel, WriteImplBase} from './write-impl-base';
+import {getSizeOfValue as defaultGetSizeOfValue} from '../get-size-of-value';
+export class LazyStore implements Store {
+  private readonly _cache: Cache;
+  private readonly _rwLock = new RWLock();
+  private readonly _baseStore: Store;
+
+  /**
+   * Store which lazily loads values from a base store and then caches them in memory.
+   *
+   * Writes only manipulate the in memory cache (thus values must be persisted to the
+   * base store through a separate process).
+   *
+   * The memory caches size is limited to `cacheSizeLimit`, and values are evicted in an LRU
+   * fashion.  `shouldBePinned` enables exempting some keys from eviction.  This mechanism
+   * should be used to prevent values not persisted to the base store from being lost.
+   * Values whose keys are pinned cannot be evicted, and their sizes are not included in the current
+   * cache size. Once a value is persisted to the base store, it's temp key for which `shouldBePinned`
+   * is `true` can be deleted, and it can be re-put with a key for which `shouldBePinned` is false (making
+   * it subject to cache eviction, which is safe since it is now persisted).
+   *
+   * @param baseStore Store to lazy load and cache values from.
+   * @param cacheSizeLimit Limit in bytes.  Size of values is determined using `getSizeOfValue`.  Keys
+   * do not count towards cache size.  Pinned values (`shouldBePinned`) do not count towards cache
+   * size.
+   * @param shouldBePinned A predicate for determining if a key should be pinned.  Keys
+   * which are not persisted to the underlying base store should be pinned, as this store does
+   * not write through to the base store (it assumes values are persisted to base store via
+   * a separate process).
+   * @param getSizeOfValue Function for measure the size in bytes of a value
+   */
+  constructor(
+    baseStore: Store,
+    cacheSizeLimit: number,
+    shouldBePinned: (key: string) => boolean,
+    getSizeOfValue = defaultGetSizeOfValue,
+  ) {
+    this._baseStore = baseStore;
+    this._cache = new Cache(
+      Math.max(cacheSizeLimit, 0),
```

Comment:
This does not seem needed. We are the consumers of this and we are not going to pass in negative numbers.

Reply:
done